[camille-readbean] iP

camille-readbean opened the pull request:
# Leto 👑

> “Once you have explored a fear, it becomes less terrifying. Part of courage comes from extending our knowledge.” -Duke Leto Atreides [source](https://www.redbubble.com/life/dune-quotes/)    

Duke Leto helps you with keeping track of todos, deadlines and events. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use
  
All you need to do is,

1.    download it from [here](https://nus-cs2103-ay2324s2.github.io/website/schedule/week4/project.html).
2.    double-click it.
3.    add your tasks.
4.    let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [X]    Managing tasks
- [X]    Managing deadlines 
- [X]    Managing events
- [X]    Extensible way to add commands 🔧 
- [X]    Save 💾  and load from csv 
- [ ]     Reminders (coming soon)


If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        letoLogo();
        TaskList.initFromFile();
        Parser.readCommandAndExecute();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

camille-readbean committed:
Level-0 rename and greet

camille-readbean committed:
Level-1 Set up echo function

camille-readbean committed:
Level-2 Add, list

camille-readbean committed:
Level 3: Mark as Done

camille-readbean committed:
Prep for next level, Todo

camille-readbean committed:
Level-4 ToDo, Event and Deadline

camille-readbean committed:
Auto text ui testing for Level-4 regression testing

camille-readbean committed:
Level 5. Handle Errors

camille-readbean committed:
Level 6. Delete
A-Enums

camille-readbean committed:
Add saving function

Current code do not support persistent tasks.

Tasks are lost on exit of program.

Add file writing and reading to csv to export and import tasks.

CSV were chosen as it can be easily imported for other purposes.

CSV format:
Type,Completed,Task,By,From,To
0   ,1        ,2   ,3 ,4   ,5

camille-readbean committed:
Merge branch 'branch-Level-7'

camille-readbean committed:
Add LocalDate to Deadline

Deadline was a String previously, giving no special meaning to it.

If deadline is a field that can capture date and have a temporal dimension we can do stuff related to time.
For example, we can give how much time remaining per task.

Change deadline from String to LocalDate
Add regex parsing instead
Update deadline saving and reading
Update Deadline creation command
Update help message

camille-readbean committed:
Add LocalDate to Event

Event from and to was a String previously, giving no special meaning to it.

If from and to is a field that can capture date and have a temporal dimension we can do stuff related to time.
For example, we can give how much time remaining per task.

Change from, to from String to LocalDate
Add regex parsing to Event
Update Event saving and reading
Update Event creation command
Update help message for Event

camille-readbean committed:
Add display for days remaining for Deadline and Event

Current code have no additional utility that make use of the benefit of dates being stored as LocalDate instead of String

New addition allows days remaining to be calculated and displayed

Add days remaining or ago to Deadline and Event
Refactor Todo to use factories
Fix some alignment issues in Ui.shortSay
Rename some factory methods to be more clear or what they are used for
Update Todo to use Regex parsing

camille-readbean committed:
Fix help message and indentation for command switch

camille-readbean committed:
Merge branch 'branch-Level-8'

camille-readbean committed:
Refactor the code to be more OOP by introducing Parser and Commands

Currently, the code is one massive switch and cases inside Parser.

Update Commands enum to make use of some Functional Interfaces so Commands are functional interfaces and can be easily changed in their respective class.

camille-readbean committed:
Merge branch 'A-MoreOOP'

ZHANGTIANYAO1 commented on Leto-Tasks.csv:
Storing data with CSV is a very good idea, learn from you. 👍

camille-readbean replied:
Thank you for your inputs. 😃    
  
I must point out that due to the use of CSV, we limit our inputs to not have commas.  
A better solution would be changing our delimiter to other characters like `|` that is usually not input by the user.  
There is another variant called `tsv` which is just using tab characters `\t` to delimit the entries.  
  
This was quite a hacky solution as the standard java library do not contain any out of the box libraries unlike Python 🐍

ZHANGTIANYAO1 commented on src/main/java/leto/parser/Parser.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    public static void readCommandAndExecute() {
+        boolean bye = false;
```
Could be `isBye` or `isExit`

camille-readbean replied:
Will fix upstream! Thanks for the lookout

ZHANGTIANYAO1 commented on src/main/java/leto/tasklist/BadTaskIndexException.java:
```diff
@@ -0,0 +1,10 @@
+public class BadTaskIndexException extends InvalidTaskException {
+    public BadTaskIndexException(int CurrentIndex) {
+        super(
+            "Bad task index number, should be a integer more than 0 and less than: " +
+            (CurrentIndex+1) +
+            " see /list for all the tasks");
```
Might be better to move `+` in front.
I noticed the same issue in several other places too

camille-readbean replied:
👷  will fix!

ZHANGTIANYAO1 commented on src/main/java/leto/tasklist/Deadline.java:
```diff
@@ -0,0 +1,71 @@
+        this.deadline = deadline;
+    }
+
+    public static Deadline DeadlineFromCMD(String input) throws InvalidTaskException {
```
Might be better to rename as  `deadlineFromCMD`
I noticed the same issue in several other places too

camille-readbean replied:
Thanks for catching that!

ZHANGTIANYAO1 commented on src/main/java/leto/tasklist/Event.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    public static Event EventFromCMD(String input) throws InvalidTaskException {
+        String regex = "(?i)event ([^,]+) /from (\\d{4}-\\d{2}-\\d{2}) /to (\\d{4}-\\d{2}-\\d{2})";
```
Wow 👍

ZHANGTIANYAO1 commented on src/main/java/leto/tasklist/TaskList.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    public static void addToList(String inputs) {
+        try {
+            Task t = null;
+            String typeOfTask = inputs.split(" ")[0];
+            switch (typeOfTask.toLowerCase()) {
+                case "event":
+                    t = Event.EventFromCMD(inputs);
+                    break;
+                case "deadline":
+                    t = Deadline.DeadlineFromCMD(inputs);
+                    break;
+                case "todo":
+                    t = Todo.TodoFromCMD(inputs);
+                    break;
```
I think there should be no indentation for case clauses

ZHANGTIANYAO1 reviewed (commented):
LGTM. 👍
Just some nits to fix.
I learned a lot.

camille-readbean committed:
Fix switch cases, naming

camille-readbean committed:
Merge branch 'add-gradle-support'

camille-readbean committed:
Amend packaging name and update build.grade

Change packaging from leto to me.ruibin.leto
Update build.grade to point to the correct package

camille-readbean committed:
Add JUnit tests for Deadline and Event

Currently, there are no unit tests.

This will introduce some tests for Deadline and Event.
Check for some invalid and valid inputs by the fromCommand and fromCsv methods.

Fix some bugs that was discovered while writing the tests for Event and Deadline as well.

camille-readbean committed:
Merge branch 'A-JUnit'

camille-readbean committed:
Add eventFromCSV_validCsvEntry_success test to Event

camille-readbean committed:
Add checkstyle to the project

camille-readbean committed:
Update code to comply with Coding Standard

Code was not complying with code standard.

Tweak code to comply with coding standard by
- renaming variables and methods
- cleaning up the ordering of imports
- updating switch statements with defaults

camille-readbean committed:
Add JavaDoc comments to non-private classes and methods.

Code did not have JavaDocs.

Add JavaDocs to code.
Edit some JavaDocs to match style.

camille-readbean committed:
Add find command to allow user to search task by keywords

Currently, can only display all the tasks with no way to narrow it down.

Add FindCommand.
Update help message.
Add FIND to Commands enum.

camille-readbean committed:
Merge branch 'A-CodingStandard'

camille-readbean committed:
Merge branch 'A-JavaDoc'

# Conflicts:
#	src/main/java/me/ruibin/leto/parser/ByeCommand.java
#	src/main/java/me/ruibin/leto/parser/DeleteCommand.java
#	src/main/java/me/ruibin/leto/parser/HelpCommand.java
#	src/main/java/me/ruibin/leto/parser/ListCommand.java
#	src/main/java/me/ruibin/leto/parser/MarkCommand.java
#	src/main/java/me/ruibin/leto/parser/NewTaskCommand.java
#	src/main/java/me/ruibin/leto/parser/SaveCommand.java
#	src/main/java/me/ruibin/leto/parser/UnmarkCommand.java
#	src/main/java/me/ruibin/leto/storage/Storage.java
#	src/main/java/me/ruibin/leto/tasklist/BadTaskIndexException.java
#	src/main/java/me/ruibin/leto/tasklist/TaskList.java

camille-readbean committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/me/ruibin/leto/ui/Ui.java

camille-readbean committed:
Fix wrong method call in FindCommand

owenong1 commented on Leto-Tasks.csv:
```diff
@@ -0,0 +1,7 @@
+D,N,20,2021-03-02,,
```
Not sure if you want to commit the data file as well. 
Maybe could specify it as a default pre-loaded list.

camille-readbean replied:
I think this was committed accidentally should be in the git ignore in the future I think!

owenong1 commented on src/main/java/me/ruibin/leto/tasklist/InvalidTaskException.java:
```diff
@@ -0,0 +1,15 @@
+import me.ruibin.leto.ui.Ui;
+
+/** Exception for invalid input for a Task */
+public class InvalidTaskException extends Exception {
```
It seems like almost all the exceptions extend InvalidTaskException; even ones like InvalidCmdException.
To me its not intuitive that an invalid command constitutes an invalid task.
Perhaps the naming/inheritance could be modified to make it more intuitive.
Otherwise, I think it makes sense!

owenong1 commented on src/main/java/me/ruibin/leto/tasklist/TaskList.java:
```diff
@@ -0,0 +1,172 @@
+public class TaskList {
+    private static final ArrayList<Task> list = new ArrayList<>(100);
+
+    public TaskList() {}
```
A bit minor but I think Java will fill in this constructor by default so you may not need this line

owenong1 commented on src/main/java/me/ruibin/leto/tasklist/TaskList.java:
```diff
@@ -0,0 +1,172 @@
+    public TaskList() {}
+
+    /** Static method to initialise the task list from csv file. */
+    public static void initFromFile() {
```
Since you used csv to refer to the file in other methods, maybe initFromCsv would be more informative

owenong1 commented on src/main/java/me/ruibin/leto/tasklist/TaskList.java:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    /** In charge of reading inputs and adding the corresponding task to the list */
+    public static void addToList(String inputs) {
```
I think you can use overloading here! Since the functionality of adding to list is the same, but the argument types are different it may not be necessary to create another named function.

camille-readbean replied:
Oh right I forgot I could have done that!

owenong1 reviewed (commented):
GJ Rui Bin!! I think your code is pretty solid overall in terms of code practices and overall logic. Only thing I can think of is to adjust some minor naming/structural aspects, but LGTM.

ATB FOR THE REST OF THE SEM!!

camille-readbean committed:
Add Varargs to TaskList addToList

Use varargs.
Simplify the adding loop in Storage.java.
As a result, removed two dependencies.

camille-readbean committed:
Merge branch 'branch-A-Varargs'

camille-readbean committed:
Add GUI

Add JavaFX support.
Refactor commands to return a Result which contains status, prints so far and the latest Exception.
Add fxml files.
Add resources for application.

camille-readbean committed:
Merge branch 'branch-Level-10'

camille-readbean committed:
Merge branch 'branch-A-CheckStyle'

camille-readbean committed:
Add some javadocs and style issue

camille-readbean committed:
Add assertions in key places

Currently, some parts of the code operate on assumptions.
For example getting a task from an ArrayList and assuming it is not null.

Adding assertions for unlikely but possible events provide more safety.

This will help us in catching errors in code.
More so if we change some of these methods that are key to the assumptions.

camille-readbean committed:
Improve code quality

Move parsing logic out of TaskList::addToString to NewTaskCommand.

Refactored some longer methods.

Make the happy path more prominent in TaskList mark and unmark method.

Rename initFromFile to initFromCsvFile to be more clear.

camille-readbean committed:
Add assertion for Control-flow invariant

Place an assertion at any location you assume will not be reached.

Add assertion for addToList for invalid task type which should not happen.

camille-readbean committed:
Modify Find to make use of Streams

FindCommand now use Stream and filter to search for the task with a keyword.

camille-readbean committed:
Merge pull request #1 from camille-readbean/branch-A-Assertions

Branch-A-Assertions

camille-readbean committed:
Fix merge conflict from Branch-A-Assertion

Amend NewTaskCommand in A-CodeQuality branch to
take into account refactoring.
A-Assertion branch added assertion to command parsing logic.
This was moved from in this branch to within the respective Command
class.

camille-readbean committed:
Merge pull request #2 from camille-readbean/branch-A-CodeQuality

merge Branch A-CodeQuality

camille-readbean committed:
Merge branch 'master' into branch-A-Streams

camille-readbean committed:
Merge pull request #3 from camille-readbean/branch-A-Streams

Merge branch-A-Streams.
Modify Find to make use of Streams.
* FindCommand now use Stream and filter to search for the task with a keyword.

camille-readbean committed:
Add workflow files for Github Actions

Create workflow file, gradle.yml.
Add gradle.yml to version tracking.
This will ensure on every push or pull request, the tests in the project
 are ran.

camille-readbean committed:
Merge branch 'branch-A-CI'

camille-readbean committed:
Add command to snooze deadlines

Add the ability to snooze the deadline of a Deadline task.
Clean up some bugs with initFromCsv.
Add extendDeadline method.
Add SnoozeCommand.
Add a getTaskByIndex to make things easier next time.
  Comes with built-in bounds checker to make sure
  it is a valid index.

camille-readbean committed:
Merge pull request #4 from camille-readbean/branch-B-Snooze

B-Snooze

camille-readbean committed:
Add changes in response to se bot

camille-readbean committed:
Add personality and update bye handling

Change bye message.
Update so bye message is displayed and disable user input.
Update references to Duke to Leto.
Update with some initial example data.
Update README.md to show gradle run.

camille-readbean committed:
Create CNAME

camille-readbean committed:
Update user guide

Update user guide.
Add Ui.png.
Updated help message in Ui.java.

camille-readbean committed:
Merge branch 'master' of https://github.com/camille-readbean/ip

camille-readbean committed:
Fix user guide

Fix emoji in user guide.
Github pages markdown do not support GitHub emojis.
